src/coreclr/vm/invokeutil.cpp

```diff
@@ -156,9 +156,15 @@ void InvokeUtil::CopyArg(TypeHandle th, PVOID argRef, ArgDestination *argDest) {
         break;
     }
 
```

Comment:
Looks like a bug, thanks for noticing. I see the arg register comes sign-extended only to the type size.

It's quite silent, though, because the upcast to int (necessary for most operations in C#) sign-extends it (shift left, then arithmetic right by 64 - typesize).

Do I need to `#ifdef` for RISC-V, or sth like this is portable enough?
```c
static_assert(sizeof(SSIZE_T) == TARGET_POINTER_SIZE);
*(SSIZE_T *)pArgDst = *(INT8 *)argRef;
```

Reply:
I think it would look best to put the code that supports sign/zero extending calling conventions under ifdef. Something like:
```
#ifdef TARGET_RISV64
   // RISCV64 calling convention requires all values to be sign or zero extended to the full register
case ELEMENT_TYPE_BOOLEAN:
case ELEMENT_TYPE_U1:
...

case ELEMENT_TYPE_U1:
...

case ELEMENT_TYPE_I2:
...

...

case ELEMENT_TYPE_R4:
...
#else

... existing code ...

#endif
```

> `static_assert(sizeof(SSIZE_T) == sizeof(ArgumentRegisters) / NUM_ARGUMENT_REGISTERS);`

The args do not have to be in registers, so this assert is not quite right. I think it is ok to omit the assert.

>  the upcast to int (necessary for most operations in C#) sign-extends it (shift left, then arithmetic right by 64 - typesize).

This sounds like an opportunity for an optimization.

Reply:
> I think it would look best to put the code that supports sign/zero extending calling conventions under ifdef. [...]

Will do, thank you.

> > the upcast to int (necessary for most operations in C#) sign-extends it (shift left, then arithmetic right by 64 - typesize).
> 
> This sounds like an opportunity for an optimization.

Indeed.